### PR TITLE
Reorder TileGroup layers to draw borders under Pixel Units

### LINKED ISSUE #6837: Feature request: Render map units on top of cultural borders
**Problem:**
Unit sprites render beneath city cultural borders, blocking units from view. See the Archer in the example below.

Example

**Proposed solution:**
Render units on top of the cultural borders so they are not blocked from view.

### PR DESCRIPTION
**Pull request by SomeTroglodyte**
Closes #6837

@Azzurite  - if you would be pleased by a little puzzle (if not ignore): See that bit with the diy bucket sort beginning at TileGroupMap L91? Begs fixing code duplication. I tried this:

<details>

```kotlin
        tileGroups.asSequence()
            .flatMap {
                sequence {
                    yield(it)
                    if (worldWrap) yieldAll(mirrorTileGroups[it.tileInfo]!!.toList())
                }
            }
            .flatMap {
                val positionalSortKey = it.tileInfo.position.run { x + y }.toInt()
                sequence<Pair<Int, Group>> {
                    // now, we steal the subgroups from all the tilegroups, that's how we form layers!
                    yield(        - positionalSortKey to it.baseLayerGroup.apply { setPosition(it.x, it.y) })
                    yield(1000000 - positionalSortKey to it.terrainFeatureLayerGroup.apply { setPosition(it.x, it.y) })
                    yield(2000000 - positionalSortKey to it.miscLayerGroup.apply { setPosition(it.x, it.y) })
                    yield(3000000 - positionalSortKey to it.pixelMilitaryUnitGroup.apply { setPosition(it.x, it.y) })
                    // pixelCivilian can share pixelMilitary 'layer', so same major sort key
                    yield(3000000 - positionalSortKey to it.pixelCivilianUnitGroup.apply { setPosition(it.x, it.y) })
                    yield(4000000 - positionalSortKey to it.highlightFogCrosshairLayerGroup.apply { setPosition(it.x, it.y) })
                    // The above layers are for the visual layers, this is for the clickability of the tile
                    yield(5000000 - positionalSortKey to it)
                    // Aaand units above everything else.
                    yield(7000000 - positionalSortKey to it.unitLayerGroup.apply { setPosition(it.x, it.y) })
                    // This is so the individual textures for the units are rendered together
                    yield(8000000 - positionalSortKey to it.unitImageLayerGroup.apply { setPosition(it.x, it.y) })
                    // city buttons + clickability
                    yield(9000000 - positionalSortKey to it.cityButtonLayerGroup.apply { setPosition(it.x, it.y) })
                }
            }
            .sortedBy { it.first }
            .forEach { this@TileGroupMap.addActor(it.second) }
```

</details>
... replacing lines 91 to 141, just before the `// there are tiles "below the zero",`. Works just fine, and on initial instrumented runs goes from ~25ms to ~6.5ms. But later runs will deteriorate to ~150ms and show significant GC activity. Yup, all those Pairs probably. (And yes the still repeated setPosition could be folded too, but I instrumented first). So - I was out of ideas how to leverage intrinsic sort stability of TimSort to simplify away those wrappers, and to get the scope of the coordinate sort back down to the original element count. Must be something simple I don't see...

### CODE REVIEW

File: core/src/com/unciv/ui/tilegroups/ActionlessGroup.kt [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+/** A [Group] with [actions], [hit] and scaling effectively disabled, pre-sized.
+ *  @param groupSize [Sets size][setSize] initially */
+open class ActionlessGroupSized(groupSize: Float) : ActionlessGroup() {
+    init {
+        isTransform = false
+        @Suppress("LeakingThis")  // works by setting fields only
+        setSize(groupSize, groupSize)
+    }
+}
```
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/ui/tilegroups/ActionlessGroup.kt:28]: Didn't think of that

File: core/src/com/unciv/ui/tilegroups/ActionlessGroup.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+/** A with [touchable] and scaling disabled, pre-sized.
+ *  @param groupSize [Sets size][setSize] initially */
+abstract class GroupSizedTouchableDisabled(groupSize: Float) : Group() {
+    init {
+        isTransform = false
+        touchable = Touchable.disabled
+        @Suppress("LeakingThis")  // works by setting fields only
+        setSize(groupSize, groupSize)
+    }
+}
```
[the remark this answers is not in the document]
Comment (SomeTroglodyte) [reply core/src/com/unciv/ui/tilegroups/ActionlessGroup.kt:39]: I probably expected reuse and wanted those "profiler names" classes to look similar... Will revisit later.

File: core/src/com/unciv/ui/map/TileGroupMap.kt [hunk 3]
```diff
@@ -149,6 +163,6 @@ class TileGroupMap<T: TileGroup>(
     fun getMirrorTiles(): HashMap<TileInfo, Pair<T, T>> = mirrorTileGroups
 
     // For debugging purposes
-    override fun draw(batch: Batch?, parentAlpha: Float) { super.draw(batch, parentAlpha) }
-    override fun act(delta: Float) { super.act(delta) }
+//     override fun draw(batch: Batch?, parentAlpha: Float) { super.draw(batch, parentAlpha) }
+//     override fun act(delta: Float) { super.act(delta) }
```
Comment (yairm210) [inline core/src/com/unciv/ui/map/TileGroupMap.kt:167]: Why would you do this... this is uncommented be cause it helps debugging... it doesn't hurt performance at all

File: core/src/com/unciv/ui/tilegroups/TileGroup.kt [hunk 4]
```diff
@@ -110,22 +108,25 @@ open class TileGroup(var tileInfo: TileInfo, val tileSetStrings:TileSetStrings,
     @Suppress("LeakingThis")    // we trust TileGroupIcons not to use our `this` in its constructor except storing it for later
     val icons = TileGroupIcons(this)
 
-    class UnitLayerGroupClass:Group(){
-        override fun draw(batch: Batch?, parentAlpha: Float) = super.draw(batch, parentAlpha)
-        override fun act(delta: Float) { // No 'snapshotting' since we trust it wil remain the same
+    class UnitLayerGroupClass(groupSize: Float) : GroupSizedTouchableDisabled(groupSize) {
+        //override fun draw(batch: Batch?, parentAlpha: Float) = super.draw(batch, parentAlpha)
```
Comment (yairm210) [inline core/src/com/unciv/ui/tilegroups/TileGroup.kt:112]: This too

File: core/src/com/unciv/ui/tilegroups/TileGroup.kt [hunk 5]
```diff
@@ -110,22 +108,25 @@ open class TileGroup(var tileInfo: TileInfo, val tileSetStrings:TileSetStrings,
...
+        override fun act(delta: Float) { // No 'snapshotting' since we trust it will remain the same
             for (child in children)
                 child.act(delta)
         }
     }
 
-    class UnitImageLayerGroupClass:ActionlessGroup(){
-        override fun draw(batch: Batch?, parentAlpha: Float) = super.draw(batch, parentAlpha)
+    class UnitImageLayerGroupClass(groupSize: Float) : ActionlessGroupSized(groupSize) {
+        //override fun draw(batch: Batch?, parentAlpha: Float) = super.draw(batch, parentAlpha)
```
Comment (yairm210) [inline core/src/com/unciv/ui/tilegroups/TileGroup.kt:120]: ...

File: core/src/com/unciv/ui/map/TileGroupMap.kt (hunk 3 above)
Comment (SomeTroglodyte) [reply core/src/com/unciv/ui/map/TileGroupMap.kt:167]: True, I doubted the "doesn't hurt performance at all" but came to no conclusive evidence.

**Comment by SomeTroglodyte** [comment]
Good suggestions, and debug overrides reinstated. One of these days I'll learn how to use them...

**Comment by letstalkaboutdune** [comment]
Thanks so much for implementing this, dev team! <3

**Pull request merged**